lib/ansible/module_utils/common/arg_spec.py

```diff
@@ -5,130 +5,266 @@
+        self._required_one_of = required_one_of
+        self._required_if = required_if
+        self._required_by = required_by
+        self.argument_spec = argument_spec
+
+    @property
+    def valid_parameter_names(self):
+        if self._valid_parameter_names:
+            return self._validated_parameter_names
+        valid_parameter_names = set()
+        for key in sorted(self.argument_spec.keys()):
+            aliases = self.argument_spec[key].get('aliases')
+            if aliases:
+                valid_parameter_names.update(["{key} ({aliases})".format(key=key, aliases=", ".join(sorted(aliases)))])
+            else:
+                valid_parameter_names.update([key])
+
+        self._valid_parameter_names = valid_parameter_names
+        return valid_parameter_names
+
+    def validate(self, parameters, *args, **kwargs):
+        """Validate module parameters against argument spec. Returns a
+        ValidationResult object.
+
+        Error messages in the ValidationResult may contain no_log values and should be
+        sanitized before logging or displaying.
 
         :Example:
 
-        validator = ArgumentSpecValidator(argument_spec, parameters)
-        passeded = validator.validate()
+        validator = ArgumentSpecValidator(argument_spec)
+        result = validator.validate(parameters)
+
+        if result.error_messages:
+            sys.exit("Validation failed: {0}".format(", ".join(result.error_messages))
+
+        valid_params = result.validated_parameters
 
         :param argument_spec: Specification of parameters, type, and valid values
         :type argument_spec: dict
 
         :param parameters: Parameters provided to the role
         :type parameters: dict
 
-        :returns: True if no errors were encountered, False if any errors were encountered.
-        :rtype: bool
+        :return: Object containing validated parameters.
+        :rtype: ValidationResult
         """
 
-        self._no_log_values.update(set_fallbacks(self.argument_spec, self._validated_parameters))
+        result = ValidationResult(parameters)
+
+        result._no_log_values.update(set_fallbacks(self.argument_spec, result._validated_parameters))
 
         alias_warnings = []
         alias_deprecations = []
         try:
-            alias_results, legal_inputs = handle_aliases(self.argument_spec, self._validated_parameters, alias_warnings, alias_deprecations)
+            alias_results, legal_inputs = _handle_aliases(self.argument_spec, result._validated_parameters, alias_warnings, alias_deprecations)
         except (TypeError, ValueError) as e:
             alias_results = {}
             legal_inputs = None
-            self._add_error(to_native(e))
+            result.errors.append(AliasError(to_native(e)))
 
         for option, alias in alias_warnings:
-            warn('Both option %s and its alias %s are set.' % (option, alias))
+            result.warnings.append({'option': option, 'alias': alias})
```

Comment:
I notice here and in `ModuleArgumentSpecValidator` that `result.warnings` seems to be specifically related to "alias warnings".  Should we differentiate that?  Or should we make it more generic?

Reply:
It should probably be more generic. It just so happens that the only warning from the validation code is related to aliases, but it would be helpful to have it be more generic.

Reply:
Maybe warnings and deprecations should be objects not just dicts? I didn't have a clear vision of how best to structure that data so it could be easily used later, so I just started with a basic dict.

Reply:
I'm on the fence about holding this up to try and come up with something clever that seems like we know what the future will hold.  With this being the only use case now, it's hard to know what another warning could look like in the future, and how we might need to represent that.